### DISCUSSION TITLE
[Question] Possible typos in lecture0 slides

### DISCUSSION THREAD

**Discussion by Allgot** [prosyslab-classroom/cs424-program-reasoning#376] [Q&A]
Hello, I found some possible typos in lecture0 slides.

1. higher-order function
In slide 18, there's an example about higher-order functions.
![image](https://user-images.githubusercontent.com/96973603/187359512-2300d5c9-b561-412e-b5b8-84ac366e81cc.png)

It says its final result will be 17, but I think it should be 18 since ((10 + 1) + 5) + 2 is 18. So I thought it was a typo, but I cannot sure because I'm totally new to OCaml.. Am I right? Or, was there any mistake in my logic?

2. "function" keyword
In slide 21, a recursive function("fact") is simplified using "function" keyword.
![image](https://user-images.githubusercontent.com/96973603/187362648-99aa43a7-fe80-41b9-87bb-2b74749d3ab0.png)

But it doesn't work and shows an error message "Error: Unbound value n".
So I think there should be "n" right after the function name, like
```
# let rec fact n = function
| 0 -> 1
| _ -> n * fact(n-1)
```
and this works fine, at least for me. Is this right?

Well I'm sorry for somewhat trivial questions, but I want to be sure about OCaml :)
Thank you for reading!

**Comment by Hhro** [comment]
Hi.

1. Yes, you are right. The result should be 18. I agree that there's a typo.

2. Does your code work? For me, a type error occurs because `fact (n - 1)` has type `int -> int`, but it is expected of type `int`.

Here is one example of the correct implementation:
```OCaml
let rec fact = function
| 0 -> 1
| n -> n * fact (n-1)
```

I think the professor mistakenly omit the `n`.

No need to be sorry. 
Thanks for reporting!

**Comment by KihongHeo** [comment]
Nice catch!
Surprising that no one has noticed them since 2020.
